### PR TITLE
UCT/CUDA/CUDA_COPY: Relaxed CUDA context dependency in cuda_copy transport.

### PR DESCRIPTION
**Pull request by rakhmets**
## What?
Relaxed CUDA context dependency in cuda_copy transport.

### CODE REVIEW

**Comment by brminich** [comment]
test failure is relevant
```
2025-03-21T19:54:57.0727290Z [ RUN      ] cuda_copy/test_p2p_no_current_cuda_ctx.get_short/1 <cuda_copy/cuda/loopback>
2025-03-21T19:54:57.0731212Z /__w/1/s/contrib/../test/gtest/common/mem_buffer.cc:419: Failure
2025-03-21T19:54:57.0732107Z cudaMemset(buffer, c, length) failed: invalid argument: ptr=0x7f521e600003 value=0 count=1
2025-03-21T19:54:57.0738553Z [  FAILED  ] cuda_copy/test_p2p_no_current_cuda_ctx.get_short/1, where GetParam() = cuda_copy/cuda/loopback (2 ms)
2025-03-21T19:54:57.0739540Z [ RUN      ] cuda_copy/test_p2p_no_current_cuda_ctx.get_zcopy/1 <cuda_copy/cuda/loopback>
2025-03-21T19:54:57.0741762Z /__w/1/s/contrib/../test/gtest/common/mem_buffer.cc:419: Failure
2025-03-21T19:54:57.0742498Z cudaMemset(buffer, c, length) failed: invalid argument: ptr=0x7f521e600003 value=0 count=1
2025-03-21T19:54:57.0876641Z [  FAILED  ] cuda_copy/test_p2p_no_current_cuda_ctx.get_zcopy/1, where GetParam() = cuda_copy/cuda/loopback (1 ms)
2025-03-21T19:54:57.0878180Z [----------] 2 tests from cuda_copy/test_p2p_no_current_cuda_ctx (3 ms total)
```

File: src/uct/cuda/cuda_copy/cuda_copy_ep.c [hunk 1]
```diff
@@ -98,26 +98,125 @@ uct_cuda_copy_get_mem_type(uct_md_h md, void *address, size_t length)
...
+    ucs_status_t status;
+    CUdevice device;
+    CUcontext ctx;
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuDeviceGetCount(&num_devices));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    for (device_num = 0; device_num < num_devices; ++device_num) {
```
Comment (yosefe) [inline src/uct/cuda/cuda_copy/cuda_copy_ep.c:115]: device_num -> device_index

File: src/uct/cuda/cuda_copy/cuda_copy_ep.c [hunk 2]
```diff
@@ -98,26 +98,125 @@ uct_cuda_copy_get_mem_type(uct_md_h md, void *address, size_t length)
...
+        }
+
+        status = uct_cuda_primary_ctx_retain(device, 0, &ctx);
+        if (status == UCS_OK) {
+            break;
+        } else if (status == UCS_ERR_NO_DEVICE) {
+            continue;
+        } else {
+            return status;
+        }
```
Comment (yosefe) [inline src/uct/cuda/cuda_copy/cuda_copy_ep.c:128]: maybe do smth like
````c
if (status == UCS_OK) {
    /* Found active context */
    status = UCT_CUDADRV_FUNC_LOG_ERR(cuCtxPushCurrent(ctx));
    if (status != UCS_OK) {
        UCT_CUDADRV_FUNC_LOG_WARN(cuDevicePrimaryCtxRelease(device));
        return status;
    }

    *cuda_device_p = device;
    *cuda_ctx_p    = ctx;
    return UCS_OK;
} else if (status != UCS_ERR_NO_DEVICE) {
    return status;
}
````

then no need to check for (device_num == num_devices) after the loop

File: src/uct/cuda/cuda_copy/cuda_copy_ep.c [hunk 3]
```diff
@@ -98,26 +98,125 @@ uct_cuda_copy_get_mem_type(uct_md_h md, void *address, size_t length)
...
-    khiter_t iter;
+    CUdevice device;
+    CUcontext ctx;
+    ucs_status_t status;
 
     result = uct_cuda_base_ctx_get_id(NULL, &ctx_id);
-    if (ucs_unlikely(result != CUDA_SUCCESS)) {
+    if (ucs_likely(result == CUDA_SUCCESS)) {
+        *ctx_id_p = ctx_id;
+        *device_p = CU_DEVICE_INVALID;
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/cuda/cuda_copy/cuda_copy_ep.c:169]: pls add a comment

File: src/uct/cuda/cuda_copy/cuda_copy_ep.c [hunk 4]
```diff
@@ -98,26 +98,125 @@ uct_cuda_copy_get_mem_type(uct_md_h md, void *address, size_t length)
...
+
+    status = uct_cuda_primary_ctx_push_first_active(&device, &ctx);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    result = uct_cuda_base_ctx_get_id(NULL, &ctx_id);
+    if (result != CUDA_SUCCESS) {
         UCT_CUDADRV_LOG(cuCtxGetId, UCS_LOG_LEVEL_ERROR, result);
+        uct_cuda_primary_ctx_pop_and_release(device);
```
Comment (yosefe) [inline src/uct/cuda/cuda_copy/cuda_copy_ep.c:181]: maybe use goto-based err flow?

File: src/uct/cuda/cuda_copy/cuda_copy_ep.c [hunk 5]
```diff
@@ -98,26 +98,125 @@ uct_cuda_copy_get_mem_type(uct_md_h md, void *address, size_t length)
...
+uct_cuda_copy_ctx_rsc_get(uct_cuda_copy_iface_t *iface, CUdevice *cuda_device_p,
+                          uct_cuda_copy_ctx_rsc_t **ctx_rsc_p)
+{
+    unsigned long long ctx_id;
+    CUdevice device;
+    ucs_status_t status;
+    khiter_t iter;
+    uct_cuda_copy_ctx_rsc_t *ctx_rsc;
+
+    status = uct_cuda_copy_ctx_try_get_id(&ctx_id, &device);
```
Comment (yosefe) [inline src/uct/cuda/cuda_copy/cuda_copy_ep.c:200]: maybe put the code of uct_cuda_copy_ctx_try_get_id here, since it seems to be used only here and both these functions are small now . so they can be united

File: src/uct/cuda/cuda_copy/cuda_copy_ep.c [hunk 6]
```diff
@@ -126,6 +225,7 @@ uct_cuda_copy_post_cuda_async_copy(uct_ep_h tl_ep, void *dst, void *src,
 {
     uct_cuda_copy_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_cuda_copy_iface_t);
     uct_base_iface_t *base_iface = ucs_derived_of(tl_ep->iface, uct_base_iface_t);
+    CUdevice device;
```
Comment (yosefe) [inline src/uct/cuda/cuda_copy/cuda_copy_ep.c:228]: let's align the code so CUdevice variables /params are named "cuda_device" everywhere.
for example, sometimes the parameter is called "CUdevice *cuda_device_p"

File: test/gtest/uct/cuda/test_switch_cuda_device.cc [hunk 7]
```diff
@@ -272,3 +272,60 @@ UCS_TEST_P(test_another_thread, detect_mem_type_cuda_fabric)
...
 
 _UCT_MD_INSTANTIATE_TEST_CASE(test_another_thread, cuda_cpy);
+
+class test_p2p_no_current_cuda_ctx : public uct_p2p_rma_test {
+public:
+    void test_xfer(send_func_t send, size_t length, unsigned flags,
+                   ucs_memory_type_t mem_type) override;
+};
+
+void test_p2p_no_current_cuda_ctx::test_xfer(send_func_t send, size_t length,
```
Comment (yosefe) [inline test/gtest/uct/cuda/test_switch_cuda_device.cc:282]: maybe rename to test_xfer_on_thread or similar?

File: test/gtest/uct/cuda/test_switch_cuda_device.cc [hunk 8]
```diff
@@ -272,3 +272,60 @@ UCS_TEST_P(test_another_thread, detect_mem_type_cuda_fabric)
...
+                                             unsigned flags,
+                                             ucs_memory_type_t mem_type)
+{
+    ucs_memory_type_t src_mem_type = UCS_MEMORY_TYPE_HOST;
+
+    mapped_buffer sendbuf(length, SEED1, sender(), 1, src_mem_type);
+    mapped_buffer recvbuf(length, SEED2, receiver(), 3, mem_type);
```
Comment (yosefe) [inline test/gtest/uct/cuda/test_switch_cuda_device.cc:289]: why need to pass offset of 1 or 3?

**Comment by yosefe** [comment]
test failures should be fixed by #10571

**Comment by brminich** [comment]
can the failure be related?
```
dcx/test_ucp_tag_mem_type.reuse_buffers_mrail/27 <dc_x,cuda_copy,rocm_copy/cuda-managed:cuda,nogdr,offload>
[     INFO ] 0 1 16 128 1048512 1048580 4194324 [swx-rdmz-ucx-gpu-02:70030:0:70030] Caught signal 11 (Segmentation fault: address not mapped to object at address 0xec451d48)
==== backtrace (tid:  70030) ====
 0  /__w/1/s/build-test/src/ucs/.libs/libucs.so.0(ucs_handle_error+0x12c) [0x7f242df2e3fc]
 1  /__w/1/s/build-test/src/ucs/.libs/libucs.so.0(+0x3c70c) [0x7f242df2e70c]
 2  /__w/1/s/build-test/src/ucs/.libs/libucs.so.0(+0x3c98b) [0x7f242df2e98b]
```

File: src/uct/cuda/cuda_copy/cuda_copy_ep.c (hunk 2 above)
Comment (rakhmets) [reply src/uct/cuda/cuda_copy/cuda_copy_ep.c]: Applied suggestion.

File: src/uct/cuda/cuda_copy/cuda_copy_ep.c (hunk 4 above)
Comment (rakhmets) [reply src/uct/cuda/cuda_copy/cuda_copy_ep.c]: Updated.

File: src/uct/cuda/cuda_copy/cuda_copy_ep.c (hunk 5 above)
Comment (rakhmets) [reply src/uct/cuda/cuda_copy/cuda_copy_ep.c]: Combined the methods.

File: test/gtest/uct/cuda/test_switch_cuda_device.cc (hunk 7 above)
Comment (rakhmets) [reply test/gtest/uct/cuda/test_switch_cuda_device.cc]: Renamed.

File: test/gtest/uct/cuda/test_switch_cuda_device.cc (hunk 8 above)
Comment (rakhmets) [reply test/gtest/uct/cuda/test_switch_cuda_device.cc]: No need, just copy/paste. Removed.

**Review by yosefe — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

**Review by brminich — APPROVED** [review]

**Comment by rakhmets** [comment]
This is a relevant failure. The issue is in the test, not in the UCT changes.
```
[ RUN      ] cuda_copy/test_p2p_no_current_cuda_ctx.get_zcopy/1 <cuda_copy/cuda/loopback>
/__w/1/s/contrib/../test/gtest/common/mem_buffer.cc:419: Failure
cudaMemset(buffer, c, length) failed: invalid argument: ptr=0x7f0f42600000 value=0 count=1
[  FAILED  ] cuda_copy/test_p2p_no_current_cuda_ctx.get_zcopy/1, where GetParam() = cuda_copy/cuda/loopback (2 ms)
```

This is another failure in `uct_dc_mlx5_ep_fence`.
```
[ RUN      ] dcx/test_ucp_fence32.atomic_add_fadd/1 <dc_x/ep_based>
[swx-rdmz-ucx-gpu-02:90739:1:91105] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x11add508)
==== backtrace (tid:  91105) ====
0  /__w/1/s/build-test/src/ucs/.libs/libucs.so.0(ucs_handle_error+0x12c) [0x7f7cd239848c]
1  /__w/1/s/build-test/src/ucs/.libs/libucs.so.0(+0x3c79c) [0x7f7cd239879c]
2  /__w/1/s/build-test/src/ucs/.libs/libucs.so.0(+0x3ca1b) [0x7f7cd2398a1b]
3  /usr/lib64/libpthread.so.0(+0xf630) [0x7f7ccdda8630]
4  /__w/1/s/build-test/src/uct/ib/mlx5/.libs/libuct_ib_mlx5.so.0(uct_dc_mlx5_ep_fence+0x19) [0x7f7cd16ae919]
5  /__w/1/s/build-test/src/ucp/.libs/libucp.so.0(+0x91508) [0x7f7cd1be5508]
6  /__w/1/s/build-test/src/ucp/.libs/libucp.so.0(ucp_atomic_op_nbx+0x21db) [0x7f7cd1beea7b]
7  /__w/1/s/build-test/test/gtest/gtest() [0xafd043]
8  /__w/1/s/build-test/test/gtest/gtest() [0xafc98c]
9  /usr/lib64/libpthread.so.0(+0x7ea5) [0x7f7ccdda0ea5]
10  /usr/lib64/libc.so.6(clone+0x6d) [0x7f7ccce7bb0d]
```

**Comment by brminich** [comment]
this failure I already saw in #10559
```
[ RUN      ] dcx/test_ucp_fence32.atomic_add_fadd/1 <dc_x/ep_based>
[swx-rdmz-ucx-gpu-02:90739:1:91105] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x11add508)
==== backtrace (tid:  91105) ====
 0  /__w/1/s/build-test/src/ucs/.libs/libucs.so.0(ucs_handle_error+0x12c) [0x7f7cd239848c]
 1  /__w/1/s/build-test/src/ucs/.libs/libucs.so.0(+0x3c79c) [0x7f7cd239879c]
 2  /__w/1/s/build-test/src/ucs/.libs/libucs.so.0(+0x3ca1b) [0x7f7cd2398a1b]
 3  /usr/lib64/libpthread.so.0(+0xf630) [0x7f7ccdda8630]
 4  /__w/1/s/build-test/src/uct/ib/mlx5/.libs/libuct_ib_mlx5.so.0(uct_dc_mlx5_ep_fence+0x19) [0x7f7cd16ae919]

```

**Comment by brminich** [comment]
but these two look related
```
[----------] 2 tests from cuda_copy/test_p2p_no_current_cuda_ctx
[ RUN      ] cuda_copy/test_p2p_no_current_cuda_ctx.get_zcopy/1 <cuda_copy/cuda/loopback>
/__w/1/s/contrib/../test/gtest/common/mem_buffer.cc:419: Failure
cudaMemset(buffer, c, length) failed: invalid argument: ptr=0x7f0f42600000 value=0 count=1
[  FAILED  ] cuda_copy/test_p2p_no_current_cuda_ctx.get_zcopy/1, where GetParam() = cuda_copy/cuda/loopback (2 ms)
[ RUN      ] cuda_copy/test_p2p_no_current_cuda_ctx.get_short/1 <cuda_copy/cuda/loopback>
/__w/1/s/contrib/../test/gtest/common/mem_buffer.cc:419: Failure
cudaMemset(buffer, c, length) failed: invalid argument: ptr=0x7f0f42600000 value=0 count=1
[  FAILED  ] cuda_copy/test_p2p_no_current_cuda_ctx.get_short/1, where GetParam() = cuda_copy/cuda/loopback (1 ms)

```

**Comment by rakhmets** [comment]
> but these two look related
> 
> ```
> [----------] 2 tests from cuda_copy/test_p2p_no_current_cuda_ctx
> [ RUN      ] cuda_copy/test_p2p_no_current_cuda_ctx.get_zcopy/1 <cuda_copy/cuda/loopback>
> /__w/1/s/contrib/../test/gtest/common/mem_buffer.cc:419: Failure
> cudaMemset(buffer, c, length) failed: invalid argument: ptr=0x7f0f42600000 value=0 count=1
> [  FAILED  ] cuda_copy/test_p2p_no_current_cuda_ctx.get_zcopy/1, where GetParam() = cuda_copy/cuda/loopback (2 ms)
> [ RUN      ] cuda_copy/test_p2p_no_current_cuda_ctx.get_short/1 <cuda_copy/cuda/loopback>
> /__w/1/s/contrib/../test/gtest/common/mem_buffer.cc:419: Failure
> cudaMemset(buffer, c, length) failed: invalid argument: ptr=0x7f0f42600000 value=0 count=1
> [  FAILED  ] cuda_copy/test_p2p_no_current_cuda_ctx.get_short/1, where GetParam() = cuda_copy/cuda/loopback (1 ms)
> ```

As I mentioned above, this is a test issue. The test executes cudaMemset in a separate thread without setting CUDA device. I will move the result check outside the function.

File: src/uct/cuda/cuda_copy/cuda_copy_ep.c [hunk 9]
```diff
@@ -187,7 +276,12 @@ uct_cuda_copy_post_cuda_async_copy(uct_ep_h tl_ep, void *dst, void *src,
     ucs_trace("cuda async issued: %p dst:%p[%s], src:%p[%s] len:%ld",
               cuda_event, dst, ucs_memory_type_names[dst_type], src,
               ucs_memory_type_names[src_type], length);
-    return UCS_INPROGRESS;
+    status = UCS_INPROGRESS;
+
+out_pop_and_release:
+    uct_cuda_primary_ctx_pop_and_release(cuda_device);
```
Comment (brminich) [inline src/uct/cuda/cuda_copy/cuda_copy_ep.c:282]: seems it is risky to release, because we are still using this context resources. If user also releases the primary context, all related resources (stream, events, etc) will be also destroyed before the operation completes.
see description in #10578

Comment (rakhmets) [reply src/uct/cuda/cuda_copy/cuda_copy_ep.c:282]: Here we retain and push only the active context. This means that the context was activated before this call. I'm not sure that we shouldn't release it here.

Comment (rakhmets) [reply src/uct/cuda/cuda_copy/cuda_copy_ep.c:282]: Also it will not help if user calls `cuDevicePrimaryCtxReset`, or `cudaDeviceReset`. It only works with user's `cuDevicePrimaryCtxRelease` call.

Moreover, we check that the context is valid before destroying the resources. And we report errors/warnings in case of failed CUDA Driver API calls. This should be enough for these cases. Maybe we can review event/stream usage and enhance error flows, and enhance unit tests by covering these corner cases.

**Review by yosefe — APPROVED** [review]

**Review by brminich — APPROVED** [review]
